sources/engine/Stride.UI/UISystem.PointerInput.cs

```diff
@@ -3,48 +3,100 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
+using Stride.Core.Diagnostics;
 using Stride.Core.Mathematics;
 using Stride.Engine;
-using Stride.Games;
 using Stride.Graphics;
 using Stride.Input;
-using Stride.UI;
+using Stride.Rendering;
+using Stride.Rendering.Compositing;
+using Stride.Rendering.UI;
 
-namespace Stride.Rendering.UI
+namespace Stride.UI
 {
-    public partial class UIRenderFeature 
+    public partial class UISystem
     {
         // object to avoid allocation at each element leave event
         private readonly HashSet<UIElement> newlySelectedElementParents = new HashSet<UIElement>();
-
         private readonly List<PointerEvent> compactedPointerEvents = new List<PointerEvent>();
-
-
         
+        private readonly HashSet<UIDocument> documents = new HashSet<UIDocument>(); 
+        
+        /// <summary>
+        /// Represents the UI-element that's currently under the mouse cursor.
+        /// Only elements with CanBeHitByUser == true are taken into account.
+        /// Last processed element_state / ?UIComponent? with a valid element will be used.
+        /// </summary>
+        public UIElement PointerOveredElement { get; internal set; }
 
-        partial void PickingUpdate(RenderUIElement renderUIElement, Viewport viewport, ref Matrix worldViewProj, GameTime drawTime, ref UIElement elementUnderMouseCursor)
-
+        private partial void UpdatePointerInput()
         {
-            if (renderUIElement.Page?.RootElement == null)
+            if (renderContext == null || sceneSystem == null || sceneSystem.GraphicsCompositor == null)
```

Comment:
> Shouldn't those checks be the callers' responsibility, or passed by the caller

Not sure, it is a private method involving the values of the class that is calling it. And if you called it else where you would want the same checks, no? I was also copying the style of `UpdateKeyEvents()` in `UISystem` which has the checks in itself as well.

>  the editor workaround is too fragile, and multi camera setup won't work appropriately because of the shared renderContext and Viewport0.

I actually agree with both of these. However, to me this feels like an issue of either me not knowing the correct way to get the required info, or the engine not exposing the info in a way that can be used. This is especially the case for `Viewport`, because as I was looking through the code to try to understand how it works and is used, there are lots of places the mention/refer to supporting multiple viewports, but not actually support it. Not sure if using Viewport is even needed, but I haven't been able to figure out exactly how and why Viewport does.
See my reply to `UIRenderFeature` for my thinking.

Reply:
Most of that would come from inheriting from `RenderFeature` and overriding `RenderFeature.Draw`, which is called by the forward renderer, scene renderer and such. 
The data you require there is setup by those systems right before calling draw. You can get your viewport from `context.CommandList.Viewport` for example.
You might be able to write a system to preemptively gather active viewports before draw, retrieve the scene system and such to take care of all of that in Update but it'll very likely look like an even larger hack than it already is

Reply:
Instead of preemptively gathering them, would probably be simple to cache a list of viewports from the last draw. Either just for the UI, or could make it more generic for the rendering system so in the future other systems could use them. I do feel like there is possibility of other gameplay/interactive use for the viewports.

Reply:
That's kind of a hack, not sure what kind of side effects this would create given that those would always be one frame behind `Update()`

Reply:
I can see how it could be a hack. I think of it more as using the most recently fully rendered data. The only rendering related data that is used is the `GraphicsDevice.Presenter.BackBuffer`'s size, and the `Viewport`. Unless the viewports change size or location there will be no idfference as they are otherwise constant. If there is a change, the only possible side effect would be the pointer position being incorrect until the next draw.

The `ForceAspectRationSceneRenderer.cs` is the only place I can see that updates a `Viewport` in a way that would impact it.

The chance of the a pointer position being off by normally a small amount for a frame seems pretty tiny to me. And worth the benefits to me.

(Of course there could be other side effects I don't see. But the code is pretty straight forward seeming to me as to what is used when. But I could be missing something!)